src/Application/Actions/Donations/Update.php

```diff
@@ -101,21 +83,27 @@ protected function action(Request $request, Response $response, array $args): Re
         }
 
         if (!isset($donationData->status)) {
-            $this->entityManager->rollback();
-
             return $this->validationError($response,
                 "Donation ID {$args['donationId']} could not be updated with missing status",
                 'New status is required'
             );
         }
 
-
+        // Lock wait errors on the select OR update mean we should try the transaction again
+        // after a short pause, up to self::MAX_UPDATE_RETRY_COUNT times.
         $retryCount = 0;
         while ($retryCount < self::MAX_UPDATE_RETRY_COUNT) {
+            $this->entityManager->beginTransaction();
+
             try {
-                if ($retryCount > 0) {
-                    $this->entityManager->refresh($donation, LockMode::PESSIMISTIC_WRITE);
+                $donation = $this->donationRepository->findAndLockOneBy(['uuid' => $args['donationId']]);
+
+                if (!$donation) {
+                    $this->entityManager->rollback();
```

Comment:
as with the deleted rollbacks, I'm confused about what we're actually rolling back, since it seems like we've only queried the DB so far, not commanded it. The only time we send a command the the DB seems to be via the call to `addData` on line 120.

Reply:
@bdsl No, we have sent a command to the DB - we have commanded it to not allow any concurrent read or write operations on this row of the donations table when we called `\MatchBot\Domain\DonationRepository::findAndLockOneBy` Rolling back the transaction unlocks the row and does make sense. I think it would happen naturally at the end of the request but it makes sense to do it early.